경북대 BE_박신영 2단계 - 회원 로그인

parknew0 opened the pull request:
안녕하세요 멘토님, 이번 STEP2 구현사항은 아래와 같고, 
선택과제였던 로그인 기능과 회원 삭제, 수정 기능은 일부만 완성해서 제출합니다!
못 다 만든 기능은 다음 STEP3와 함께 완성해볼까 합니다!

회원 가입 및 로그인 API 구현
- JWT를 사용한 토큰 기반 인증 시스템을 구현했습니다.
- /api/members/register 와 /api/members/login 엔드포인트를 통해 기능을 제공합니다.

토큰 인증 인터셉터 구현
- HandlerInterceptor를 사용하여 Authorization 헤더의 토큰을 검증하는 인증 로직을 추가했습니다.
- 로그인/회원가입 등 일부 경로를 제외한 API 요청은 유효한 토큰이 필요합니다.

회원 관리자 페이지 UI 기본 구현
- Thymeleaf를 사용하여 관리자가 회원을 관리할 수 있는 UI를 구현했습니다.
- 아직까지는 현재 회원 목록 조회 및 새 회원 등록 기능까지만 구현되어 있습니다.

기본 내비게이션 추가
- 초기 화면('/')에서 각 관리자 페이지로 이동할 수 있는 링크를 추가하여 사용자 편의성을 개선했습니다.

추후 작업
- 회원 관리자 페이지의 수정 및 삭제 기능
- 일반 사용자를 위한 로그인 폼 UI

parknew0 committed:
refactor(exception): simplify global exception handler

parknew0 committed:
feat(docs): readme.md STEP2 구현 기능 목록 추가

parknew0 committed:
feat: add jjwt library, JWT 사용을 위한 라이브러리 추가

parknew0 committed:
feat(member): member entity 와 repository 추가

parknew0 committed:
feat(member): member registration 구현

parknew0 committed:
feat(docs): read.me 파일에 member registration 구현부분 마저 반영

parknew0 committed:
feat(member): 로그인 엔드포인트 추가 및 서비스 뼈대 구축

parknew0 committed:
feat(member): member 로그인 및 토큰 발급 구현

parknew0 committed:
feat(auth): token validation logic을 jwtprovider에 추가

parknew0 committed:
feat(auth): implement auth interceptor to verify jwt

parknew0 committed:
feat(admin): 어드민 페이지에 'find all members' 기능 추가

parknew0 committed:
feat(admin): 새로운 멤버 추가하는 페이지와 로직을 추가

parknew0 committed:
feat(ui): 홈페이지 추가하고, 네비게이션 링크도 추가

rayfault commented on README.md:
```diff
@@ -14,3 +14,20 @@
 - [x] REST API를 위한 글로벌 예외 처리기 (`@RestControllerAdvice`) 구현
 - [x] 관리자 페이지 폼 유효성 검사 실패 처리
+
+## 구현할 기능 목록 (2단계 - 회원 로그인)
```
👍🏻

rayfault commented on src/main/java/gift/auth/AuthInterceptor.java:
```diff
@@ -0,0 +1,36 @@
+            throw new AuthenticationException("유효하지 않은 토큰입니다.");
+        }
+
+        String token = authHeader.substring(7);
```
매직넘버 7 과 "Bearer " 를 개선해 볼 수 있을까요?

parknew0 replied:
"Bearer "를 상수로 변경하고, 그 상수의 길이를 가져오도록 수정해서 더 깔끔해진 듯 합니다!

rayfault commented on src/main/java/gift/auth/JwtTokenProvider.java:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    public String getPayload(String token) {
```
메소드명이 `getSubject` 여야 하지는 않을까요? 동작과 달라보입니다.

parknew0 replied:
좀 더 세부적으로 이름을 나타내는 편이 좋았었네요! 감사합니다.

rayfault commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @PostMapping("/register")
+    public ResponseEntity<MemberResponse> register(@RequestBody MemberRequest request) {
```
토큰의 응답이라는 이름을 좀 더 나타내 볼 수 있을까요?
공용으로 사용되는 MemberResponse 라 실제로 들어있는 데이터를 생각해보기가 어려울 것 같아요.

parknew0 replied:
MemberResponse 안에 token만 들어있어서, 아예 TokenResponse 라는 이름으로 수정했습니다!

rayfault commented on src/main/java/gift/exception/AuthenticationException.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.UNAUTHORIZED)
```
이 설정은 없어도 되지 않을까요?

parknew0 replied:
GlobalExceptionHandler의 @ExceptionHandler가 더 높은 우선순위를 갖기 때문에 @ResponseStatus는 쓰이지조차 않네요, 감사합니다!

rayfault commented on src/main/java/gift/repository/JdbcMemberRepository.java:
```diff
@@ -0,0 +1,79 @@
+        }, keyHolder);
+
+        Long generatedId = keyHolder.getKey().longValue();
+        return member.assignId(generatedId);
```
👍🏻

rayfault commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public MemberResponse register(MemberRequest request) {
+        Member member = new Member(request.email(), request.password());
```
데이터베이스에 저장할 때 비밀번호를 사용자의 요청 그대로 저장하게 되지 않을까요?
암호화에 대한 명시적인 요청은 없지만, 암호화를 한 번 해보면 어떨까요? PasswordEncoder 등을 통해 쉽게 해보실 수 있을거라 생각합니다~

parknew0 replied:
좋은 인사이트 감사합니다! 안내해주신 방법 적용했습니다. 이 방법을 적용하면 데이터 베이스가 유출되더라도 실제 비밀번호는 보호되네요

rayfault commented on src/main/java/gift/repository/JdbcMemberRepository.java:
```diff
@@ -0,0 +1,79 @@
+public class JdbcMemberRepository implements MemberRepository {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final RowMapper<Member> memberRowMapper = (rs, rowNum) -> new Member(
```
SpringFramework 에서 제공하는 RowMapper 구현체를 사용해 볼 수 있을까요?

parknew0 replied:
리팩토링을 진행했습니다!

rayfault commented on src/main/java/gift/auth/AuthInterceptor.java:
```diff
@@ -0,0 +1,36 @@
+            throw new AuthenticationException("유효하지 않은 토큰입니다.");
+        }
+
+        request.setAttribute("userEmail", jwtTokenProvider.getPayload(token));
```
jwt 에 많은 내용을 담을 수 있긴 하지만, userEmail 보다는 id 를 사용하는 것은 어떨까요?
여러가지 이유가 있지만, 가장 단순하게는 email 보다 보안이 좋고 변경의 여지가 적습니다.

parknew0 replied:
관련된 코드 모두 수정완료했습니다! id는 데이터베이스가 자동으로 부여하는 숫자라서, 이메일보다 더 개인을 식별하기 힘들다는 점에서, 더 보안을 신경 토큰 사용방식이라는 것으로 이해했습니다!

rayfault commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,54 @@
+                .toList();
+    }
+
+    public void saveMember(MemberRequest request) {
```
이미 존재하는 사용자의 경우 어떻게 처리될까요?
예외 처리가 필요해보입니다 ^^

parknew0 replied:
현재는 이메일 중복을 막고있지 않아서 중복된 사용자가 존재할 수 있다는 문제가 있네요! 수정했습니다.

rayfault commented on src/main/java/gift/controller/admin/AdminMemberController.java:
```diff
@@ -0,0 +1,42 @@
+
+@Controller
+@RequestMapping("/admin/members")
+public class AdminMemberController {
```
이 화면의 경우 인증되고, 권한이 있는 사용자만 처리를 해야 할 것으로 보이는데요. 어떻게 처리될 수 있을까요?
지금으로서는 모두가 접근할 수 있는 페이지로 보입니다~

parknew0 replied:
WebConfig에서 /admin/** 경로에 AuthInterceptor를 적용하도록 설정했습니다. 
AuthInterceptor는 해당 경로로 요청이 올 때, 쿠키나 헤더에 유효한 JWT 토큰이 없으면 /admin/login 페이지로 리다이렉트시킵니다.
로그인하지 않은 사용자(인증되지 않은 사용자)는 절대 /admin/members 페이지에 접근할 수 없도록 수정했습니다!

rayfault reviewed (changes requested):
어드민 기능에 대해서 잘 구현된 토큰을 이용해보는 것은 어떨까요? 구현해 주신 내용을 보면 충분히 적용이 가능하리라 생각합니다. 그 외에 몇가지 코멘트를 달았는데요. 함께 확인 부탁드립니다.

1차 리뷰 마치겠습니다. 고생하셨습니다.

parknew0 committed:
feat(auth): 관리자 로그인 기능 추가

보안 강화를 위해 관리자 페이지에 토큰 기반 인증 시스템을 구현. 기존 관리자 섹션은 누구나 접근할 수 있다는 문제가 있었음

- 관리자 전용 로그인 인터페이스를 제공하기 위해 `AdminAuthController`와 `login-form.html`을 추가
- 로그인 성공 시 JWT를 발급하고, 안전한 `HttpOnly` 쿠키에 저장하여 브라우저 기반의 세션 관리에 적합하도록 구현
- `AuthInterceptor`가 헤더 및 쿠키 기반 토큰을 모두 지원하도록 수정하여, API 클라이언트와 웹 어드민 세션을 단일 인터셉터로 처리
- `WebConfig`를 수정하여 모든 `/admin/**` 경로를 보호하고, 로그인 페이지 자체에 대한 접근을 허용하기 위해 `/admin/login` 경로는 제외

parknew0 committed:
refactor(auth): 매직넘버 및 하드코딩 문자열 개선

parknew0 committed:
refactor(JwtTokenProvider): getPayload를 getSubject로 변경

parknew0 committed:
refactor(dto): MemberResponse를 TokenResponse로 이름 변경

parknew0 committed:
refactor(exception): remove unnecessary ResponseStatus annotation from AuthenticationException

parknew0 committed:
feat(security): 회원 비밀번호 암호화 기능 구현

parknew0 committed:
refactor(repository): replace custom RowMapper with BeanPropertyRowMapper for Member entity

parknew0 commented:
현재 프로젝트에는 스프링 시큐리티의 SecurityFilterChain과 우리가 직접 만든 AuthInterceptor라는 두 개의 보안 시스템이 함께 동작하고 있습니다.

즉, 어떤 경로는 허용하고 어떤 경로는 막을지에 대한 설정이 SecurityConfig와 WebConfig 두 군데에 나뉘어 있습니다. 예를 들어 새로운 공개 API를 추가할 때, 두 파일 모두를 수정해야 할 수도 있어 실수의 여지가 있을 것 같습니다.

1차 검문소 역할은 스프링 시큐리티(SecurityFilterChain)가, 2차 신원 확인소 역할은 AuthInterceptor 인터셉터가 하는 모양새인데, 이걸 하나로 통합을 하는 것이 나은지, 아니면 이렇게 역할을 분리해서 두는게 좋은지 고민 중입니다.

parknew0 committed:
feat(security): security filter chain 설정을 통해 API 엔드포인트 인증 필요유무 세팅

parknew0 committed:
refactor(auth): JWT 식별자를 이메일에서 사용자 ID로 변경

parknew0 committed:
fix(member): 회원 가입 시 이메일 중복 예외 처리

parknew0 commented:
현재 코드는 아래와 같은 문제가 있습니다.

1. 서버를 껐다 켤 때마다 H2 인메모리 데이터베이스는 초기화되어 모든 회원 정보가 사라집니다.

2. 하지만 크롬 웹 브라우저에는 이전에 로그인 성공 시 발급받았던 쿠키(토큰)가 그대로 남아있었습니다.

3. AuthInterceptor는 이 토큰의 서명과 유효기간이 올바르다는 것까지만 확인했기 때문에, 데이터베이스에 실제 사용자가 없는데도 요청을 통과시켰던 것입니다.

이 부분도 추후 개선하면 좋을것같습니다.

parknew0 committed:
fix(security): 보안 접근규칙 수정

rayfault commented:
> 현재 프로젝트에는 스프링 시큐리티의 SecurityFilterChain과 우리가 직접 만든 AuthInterceptor라는 두 개의 보안 시스템이 함께 동작하고 있습니다.


지금으로서는 `.anyRequest().permitAll()` 로 모든 요청을 허용하는 것으로 보이는데요.
그렇기 때문에 securityFilterChain 설정은 없어도 될 것으로 보입니다. 특히나 이번 과제의 경우 Interceptor 를 통해 과제 구현을 목표로 삼고 있기 때문에 이 부분은 삭제해주는게 더 깔끔해보입니다~

rayfault commented on src/main/java/gift/auth/AuthInterceptor.java:
```diff
@@ -0,0 +1,65 @@
+            throw new AuthenticationException("유효하지 않은 토큰입니다.");
+        }
+
+        request.setAttribute("userId", Long.parseLong(jwtTokenProvider.getSubject(token)));
```
사용자를 `member` 로 잡고 있어서 `memberId` 가 더 맞는 이름일 것 같군요!

rayfault commented on src/main/java/gift/entity/Member.java:
```diff
@@ -0,0 +1,50 @@
+
+    }
+
+    public Member(Long id, String email, String password) {
```
생성자 구현을 모두 정적 팩토리 메소드로 바꾸면, 기본 생성자도 선언하지 않을 수 있겠네요!

rayfault reviewed (approved):
Spring Security 까지 살짝 적용된 것 보았는데요, 별도의 코멘트를 남겨두었으니 확인 후 다음 과제 때 적용해주시면 좋겠습니다.

개선이 이루어지면 좋겠다 하는 부분 모두 잘 변경해주셨네요. 리뷰 반영 감사합니다!
리뷰 마치겠습니다. 고생하셨습니다~ 🙇🏻‍♂️
